File: app/src/main/java/org/bobstuff/bobboardview/app/trello/BoardActivity.java
package org.bobstuff.bobboardview.app.trello;

import android.animation.Animator;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearSnapHelper;
import android.support.v7.widget.SnapHelper;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.DragEvent;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.Window;
import android.view.WindowManager;
import android.widget.TextView;

import org.bobstuff.bobboardview.BobBoardListAdapter;
import org.bobstuff.bobboardview.R;
import org.bobstuff.bobboardview.BobBoardView;
import org.bobstuff.bobboardview.BobBoardAdapter;
import org.bobstuff.bobboardview.app.trello.model.Board;
import org.bobstuff.bobboardview.app.trello.model.BoardList;
import org.bobstuff.bobboardview.app.trello.model.Card;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

public class BoardActivity extends AppCompatActivity implements BobBoardView.BobBoardViewListener {
    private TrelloBoardAdapter mTrelloBoardAdapter;
    private Toolbar toolbar;
    private TextView te3;
    private BobBoardView mBobBoardView;
    private Board board;
    private SnapHelper mSnapHelper;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_board);

        board = createTestData();

        mTrelloBoardAdapter = new TrelloBoardAdapter(getBaseContext(), getResources().getDimensionPixelSize(R.dimen.trello_item_list_width));
        mTrelloBoardAdapter.setLists(board.getLists());

        mBobBoardView = findViewById(R.id.board_view);
        mBobBoardView.setBoardAdapter(mTrelloBoardAdapter);
        mBobBoardView.setBoardViewListener(this);
        CustomDividers.ListItemDecoration dividerItemDecoration = new CustomDividers.ListItemDecoration(40);
        mBobBoardView.getListRecyclerView().addItemDecoration(dividerItemDecoration);

        mSnapHelper = new LinearSnapHelper();
        mSnapHelper.attachToRecyclerView(mBobBoardView.getListRecyclerView());

        toolbar = (Toolbar) findViewById(R.id.my_toolbar);
        setSupportActionBar(toolbar);
        toolbar.setBackgroundDrawable(new ColorDrawable(Color.parseColor("#00407f")));
        toolbar.setTitleTextColor(Color.parseColor("#ffffff"));
        Drawable upArrow = getResources().getDrawable(R.drawable.abc_ic_ab_back_material);
        upArrow.setColorFilter(Color.parseColor("#ffffff"), PorterDuff.Mode.SRC_ATOP);
        getSupportActionBar().setHomeAsUpIndicator(upArrow);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        setStatusBarColor();

        te3 = (TextView) findViewById(R.id.stupid_text);
        te3.setVisibility(View.INVISIBLE);
        te3.setOnDragListener(new View.OnDragListener() {
            @Override
            public boolean onDrag(View v, DragEvent event) {
                final int action = event.getAction();

                // Handles each of the expected events
                switch(action) {
                    case DragEvent.ACTION_DRAG_ENTERED:
                        Log.d("TEST", "ENTERED TOP BAR");
                        toolbar.setBackgroundDrawable(new ColorDrawable(Color.parseColor("#00ff00")));
                        break;
                    case DragEvent.ACTION_DRAG_EXITED:
                        Log.d("TEST", "EXITED TOP BAR");
                        toolbar.setBackgroundDrawable(new ColorDrawable(Color.parseColor("#00407f")));
                        break;
                }

                return true;
            }
        });
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

    public void onDragStarted(BobBoardView.DragType dragType) {
        mSnapHelper.attachToRecyclerView(null);
        if (dragType == BobBoardView.DragType.LIST) {
            ActionBar mActionBar = getSupportActionBar();
            mActionBar.setDisplayShowHomeEnabled(false);
            mActionBar.setDisplayShowTitleEnabled(false);
            // previously invisible view
            // get the center for the clipping circle
            int cx = te3.getWidth() / 2;
            int cy = te3.getHeight() / 2;
            // get the final radius for the clipping circle
            float finalRadius = (float) Math.hypot(cx, cy);
            // create the animator for this view (the start radius is zero)
            Animator anim =
                    ViewAnimationUtils.createCircularReveal(te3, cx, cy, 0, finalRadius);
            // make the view visible and start the animation
            te3.setVisibility(View.VISIBLE);
            anim.start();

            mTrelloBoardAdapter.triggerScaleDownAnimationForListDrag();
        }
    }

    public void onDragEnded(BobBoardView.DragType dragType) {
        mSnapHelper.attachToRecyclerView(mBobBoardView.getListRecyclerView());
        if (dragType == BobBoardView.DragType.LIST) {
            int cx = te3.getWidth() / 2;
            int cy = te3.getHeight() / 2;
            // get the final radius for the clipping circle
            float finalRadius = (float) Math.hypot(cx, cy);
            Animator anim =
                    ViewAnimationUtils.createCircularReveal(te3, cx, cy, 0, finalRadius);
            anim.addListener(new Animator.AnimatorListener() {
                @Override
                public void onAnimationStart(Animator animation) {
                }

                @Override
                public void onAnimationEnd(Animator animation) {
                    te3.setVisibility(View.INVISIBLE);
                    ActionBar mActionBar = getSupportActionBar();
                    mActionBar.setDisplayShowHomeEnabled(true);
                    mActionBar.setDisplayShowTitleEnabled(true);
                }

                @Override
                public void onAnimationCancel(Animator animation) {
                }

                @Override
                public void onAnimationRepeat(Animator animation) {
                }
            });
            // make the view visible and start the animation
            anim.start();

            mTrelloBoardAdapter.triggerScaleUpAnimationForListDrag();
        }
    }

    public void onListMove(int fromPosition, int toPosition) {
        Log.d("TEST", "List moved");
        mTrelloBoardAdapter.onItemMove(fromPosition, toPosition);
        Collections.swap(board.getLists(), fromPosition, toPosition);
    }

    public void onCardMove(BobBoardAdapter.ListViewHolder listViewHolder, int listIndex, int fromPosition, int toPosition) {
        Log.d("TEST", "card moved");
        ((TrelloBoardAdapter.TrelloListViewHolder)listViewHolder).mTrelloListAdapter.onItemMove(fromPosition, toPosition);

        Collections.swap(board.getLists().get(listIndex).getCards(), fromPosition, toPosition);
    }

    public void onCardMoveList(BobBoardAdapter.ListViewHolder toViewHolder, int toIndex) {
        //Card card = ((TrelloBoardAdapter.TrelloListViewHolder)fromViewHolder).dismissDraggedCard(fromIndex);
        //((TrelloBoardAdapter.TrelloListViewHolder)toViewHolder).add(toIndex, card);

        //Card card2 = board.getLists().get(fromViewHolder.getAdapterPosition()).getCards().remove(fromIndex);
        //board.getLists().get(toViewHolder.getAdapterPosition()).getCards().add(toIndex, card2);
    }

    public Board createTestData() {
        Card cardOne = new Card("Infinite Scrolling on main collection view, Infinite Scrolling on main collection view, Infinite Scrolling on main collection view");
        Card cardTwo = new Card("Check synchronous image load of separate thread isn't bad");
        Card cardThree = new Card("Independent controller logic");
        Card cardFour = new Card("Allow sorting by dragging and dropping of cards");
        Card cardFive = new Card("Sent delta change for hand and counts to opponents");
        Card cardSix = new Card("double tap to play should be disabled when it is not your turn to play");
        Card cardSeven = new Card("Add room support");

        BoardList listOne = new BoardList("todo", new ArrayList<>(Arrays.asList(cardOne, cardTwo)));
        BoardList listTwo = new BoardList("in progress", new ArrayList<>(Arrays.asList(cardThree, cardFour, cardFive)));
        BoardList listThree = new BoardList("blocked");
        BoardList listFour = new BoardList("testing");
        BoardList listFive = new BoardList("test failed", new ArrayList<>(Arrays.asList(cardSix)));
        BoardList listSix = new BoardList("complete", new ArrayList<>(Arrays.asList(cardSeven)));

        return new Board(new ArrayList<>(Arrays.asList(listOne, listTwo, listThree, listFour, listFive, listSix)));
    }

    public void setStatusBarColor(){
        Window window = this.getWindow();
        window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.setStatusBarColor(getDarkerShadeColor(Color.parseColor("#00407f")));
        }
    }

    /**
     * https://blog.fossasia.org/creating-different-shades-of-a-color-in-phimpme-android/
     * @param c
     * @return
     */
    public int getDarkerShadeColor(int c){
        float[] hsv = new float[3];
        int color = c;
        Color.colorToHSV(color, hsv);
        hsv[2] *= 0.80f;
        color = Color.HSVToColor(hsv);
        return color;
    }

    public void onStartDrag() {

    }

    @Override
    public void onListDragStarted(@NotNull BobBoardAdapter.ListViewHolder<? extends BobBoardListAdapter<?>> listViewHolder) {

    }

    @Override
    public void onCardDragStarted(@NotNull BobBoardAdapter.ListViewHolder<? extends BobBoardListAdapter<?>> listViewHolder, @NotNull BobBoardListAdapter.CardViewHolder cardViewHolder) {

    }

    @Override
    public void onCardDragExitedList(@NotNull BobBoardAdapter.ListViewHolder<? extends BobBoardListAdapter<?>> listViewHolder, @NotNull BobBoardListAdapter.CardViewHolder cardViewHolder) {

    }

    @Override
    public void onCardDragEnteredList(@NotNull BobBoardAdapter.ListViewHolder<? extends BobBoardListAdapter<?>> listViewHolder, @NotNull BobBoardListAdapter.CardViewHolder cardViewHolder) {

    }

    @Override
    public void onListDragEnded(@Nullable BobBoardAdapter.ListViewHolder<? extends BobBoardListAdapter<?>> listViewHolder) {

    }

    @Override
    public void onCardDragEnded(@Nullable BobBoardAdapter.ListViewHolder<? extends BobBoardListAdapter<?>> listViewHolder, @Nullable BobBoardListAdapter.CardViewHolder cardViewHolder) {

    }
}

File: app/src/main/java/org/bobstuff/bobboardview/app/trello/BoardActivity.kt
package org.bobstuff.bobboardview.app.trello

import android.animation.Animator
import android.graphics.Color
import android.graphics.PointF
import android.graphics.Rect
import android.graphics.drawable.ColorDrawable
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.support.v7.widget.*
import android.util.Log
import android.util.TypedValue
import android.view.DragEvent
import android.view.MenuItem
import android.view.View
import android.view.ViewAnimationUtils
import android.widget.TextView
import android.widget.Toast
import org.bobstuff.bobboardview.*
import org.bobstuff.bobboardview.app.R
import org.bobstuff.bobboardview.app.scrum.ScrumColumnAdapter
import org.bobstuff.bobboardview.app.trello.model.BoardList

import org.bobstuff.bobboardview.app.trello.model.Card
import org.bobstuff.bobboardview.app.util.DragOperation
import org.bobstuff.bobboardview.app.util.getWindowSize
import org.bobstuff.bobboardview.app.util.setStatusBarColor
import org.bobstuff.bobboardview.app.util.styleToolbarContent

import java.util.Collections

class BoardActivity : AppCompatActivity() {
    private lateinit var trelloBoardAdapter: TrelloBoardAdapter
    private lateinit var toolbar: Toolbar
    private lateinit var boardView: BobBoardView
    private var snapHelper: SnapHelper? = null
    private lateinit var archiveView: TextView
    private var board = createTestData()
    private val currentDragOperation = BobBoardDragOperation<Card, BoardList>()
    private var dragOperationInsideArchiveView: Boolean = false
    private var dragOperationEnteredArchiveView: Boolean = false
    private var dragOperationPositionInArchiveView = PointF()
    private var largeScreen: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_board)

        trelloBoardAdapter = TrelloBoardAdapter(this, currentDragOperation, resources.getDimensionPixelSize(R.dimen.trello_item_list_width))
        trelloBoardAdapter.setItems(board.lists)

        val config = SimpleBobBoardViewListener.SimpleBobBoardViewListenerConfigBuilder()
                .removeItemWhenDraggingBetweenLists(true)
                .removeListWhenDraggingOutsideBoardView(true)
                .build()
        val listener = CustomBobBoardViewListener(currentDragOperation, config)

        boardView = findViewById(R.id.board_view)
        boardView.boardAdapter = trelloBoardAdapter
        boardView.setBoardViewListener(listener)
        boardView.listRecyclerView.contentDescription = "column recycler"
        boardView.listRecyclerView.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)

        val dividerSizeInPixels = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 20f, resources.displayMetrics)
        val dividerItemDecoration = BobBoardSimpleDividers(dividerSizeInPixels.toInt(), BobBoardSimpleDividersOrientation.HORIZONTAL)
        boardView.listRecyclerView.addItemDecoration(dividerItemDecoration)

        val windowSize = getWindowSize()
        val listSize = resources.getDimensionPixelSize(R.dimen.trello_item_list_width)
        if ((windowSize.x / listSize) < 2) {
            snapHelper = LinearSnapHelper()
        }
        if ((windowSize.x / listSize) > 3) {
            largeScreen = true
        }
        snapHelper?.attachToRecyclerView(boardView.listRecyclerView)

        toolbar = findViewById(R.id.my_toolbar)
        setSupportActionBar(toolbar)
        styleToolbarContent(toolbar, Color.parseColor("#ffffff"))
        setStatusBarColor(Color.parseColor("#003365"), false)

        archiveView = findViewById(R.id.archive_view)
        archiveView.visibility = View.INVISIBLE
        archiveView.setOnDragListener { _, event ->
            val action = event.action
            when (action) {
                DragEvent.ACTION_DRAG_ENTERED -> {
                    dragOperationEnteredArchiveView = true
                }
                DragEvent.ACTION_DRAG_LOCATION -> {
                    dragOperationPositionInArchiveView.set(event.x, event.y)
                    if (dragOperationEnteredArchiveView) {
                        dragOperationEnteredArchiveView = false
                        dragOperationInsideArchiveView = true
                        toolbar.background = ColorDrawable(Color.parseColor("#00ff00"))
                        if (currentDragOperation.dragType == BobBoardView.DragType.LIST) {
                            removeDragListItem()
                        }
                    }
                }
                DragEvent.ACTION_DRAG_EXITED -> {
                    dragOperationEnteredArchiveView = false
                    if (dragOperationInsideArchiveView) {
                        dragOperationInsideArchiveView = false
                        toolbar.background = ColorDrawable(Color.parseColor("#00407f"))
                        if (currentDragOperation.dragType == BobBoardView.DragType.LIST) {
                            insertDragListItem(dragOperationPositionInArchiveView.x, dragOperationPositionInArchiveView.y)
                        }
                    }
                }
            }
            true
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }

    inner class CustomBobBoardViewListener(currentDragOperation: BobBoardDragOperation<Card, BoardList>, config: SimpleBobBoardViewListener.SimpleBobBoardViewListenerConfig):
            SimpleBobBoardViewListener<Card, BoardList>(currentDragOperation, config) {

        override fun removeCardFromDataSource(listIndex: Int, cardIndex: Int): Card {
            return board.lists[listIndex].cards.removeAt(cardIndex)
        }

        override fun addListToDataSource(listIndex: Int, boardList: BoardList) {
            board.lists.add(listIndex, boardList)
        }

        override fun addCardToDataSource(listIndex: Int, cardIndex: Int, card: Card) {
            board.lists[listIndex].cards.add(cardIndex, card)
        }

        override fun moveCardInDataSource(listIndex: Int, fromCardIndex: Int, toCardIndex: Int) {
            Collections.swap(board.lists[listIndex].cards, fromCardIndex, toCardIndex)
        }

        override fun moveListInDataSource(fromListIndex: Int, toListIndex: Int) {
            Collections.swap(board.lists, fromListIndex, toListIndex)
        }

        override fun removeListFromDataSource(listIndex: Int): BoardList {
            return board.lists.removeAt(listIndex)
        }

        override fun onCardDragEnded(boardView: BobBoardView, activeListViewHolder: BobBoardAdapter.ListViewHolder<BobBoardListAdapter<*>>?,
                                     listViewHolder: BobBoardAdapter.ListViewHolder<BobBoardListAdapter<*>>?,
                                     cardViewHolder: BobBoardListAdapter.CardViewHolder?) {
            if (!dragOperationInsideArchiveView) {
                Toast.makeText(this@BoardActivity, "Card moved from col: ${currentDragOperation.startingListIndex}, " +
                        "row: ${currentDragOperation.startingCardIndex} to col: ${currentDragOperation.listIndex}, " +
                        "row: ${currentDragOperation.cardIndex}", Toast.LENGTH_LONG).show()
            }
            if (dragOperationInsideArchiveView && currentDragOperation.orphaned) {
                Toast.makeText(this@BoardActivity, "Card archived", Toast.LENGTH_LONG).show()
                currentDragOperation.orphaned = false
            }

            val finalColumnPosition = currentDragOperation.listIndex

            super.onCardDragEnded(boardView, activeListViewHolder, listViewHolder, cardViewHolder)

            cardViewHolder?.itemView?.visibility = View.VISIBLE

            val anim = centeredCircleAnimation(archiveView, true)
            anim.addListener(object : Animator.AnimatorListener {
                override fun onAnimationStart(animation: Animator) {}

                override fun onAnimationEnd(animation: Animator) {
                    archiveView.visibility = View.INVISIBLE
                    toolbar.background = ColorDrawable(Color.parseColor("#00407f"))
                    val actionBar = supportActionBar!!
                    actionBar.setDisplayShowHomeEnabled(true)
                    actionBar.setDisplayShowTitleEnabled(true)
                    snapHelper?.attachToRecyclerView(boardView.listRecyclerView)
                }

                override fun onAnimationCancel(animation: Animator) {}

                override fun onAnimationRepeat(animation: Animator) {}
            })
            anim.start()

            boardView.listRecyclerView.smoothScrollToPosition(finalColumnPosition)
        }

        override fun onListDragStarted(boardView: BobBoardView, listViewHolder: BobBoardAdapter.ListViewHolder<BobBoardListAdapter<*>>) {
            super.onListDragStarted(boardView, listViewHolder)

            snapHelper?.attachToRecyclerView(null)
            val actionBar = supportActionBar!!
            actionBar.setDisplayShowHomeEnabled(false)
            actionBar.setDisplayShowTitleEnabled(false)

            val anim = centeredCircleAnimation(archiveView, false)
            archiveView.visibility = View.VISIBLE
            anim.start()

            trelloBoardAdapter.triggerScaleDownAnimationForListDrag(listViewHolder)
        }

        override fun onCardDragStarted(boardView: BobBoardView, listViewHolder: BobBoardAdapter.ListViewHolder<BobBoardListAdapter<*>>, cardViewHolder: BobBoardListAdapter.CardViewHolder) {
            super.onCardDragStarted(boardView, listViewHolder, cardViewHolder)

            snapHelper?.attachToRecyclerView(null)

            val actionBar = supportActionBar!!
            actionBar.setDisplayShowHomeEnabled(false)
            actionBar.setDisplayShowTitleEnabled(false)

            val anim = centeredCircleAnimation(archiveView, false)
            archiveView.visibility = View.VISIBLE
            anim.start()
        }

        override fun onListDragEnded(boardView: BobBoardView, listViewHolder: BobBoardAdapter.ListViewHolder<BobBoardListAdapter<*>>?) {
            Log.d("TEST", "DID THIS THINGY GET CALLED")
            if (dragOperationInsideArchiveView && currentDragOperation.orphaned) {
                currentDragOperation.orphaned = false
            }

            super.onListDragEnded(boardView, listViewHolder)
            listViewHolder?.itemView?.visibility = View.VISIBLE

            val anim = centeredCircleAnimation(archiveView, true)
            anim.addListener(object : Animator.AnimatorListener {
                override fun onAnimationStart(animation: Animator) {}

                override fun onAnimationEnd(animation: Animator) {
                    archiveView.visibility = View.INVISIBLE
                    val actionBar = supportActionBar!!
                    actionBar.setDisplayShowHomeEnabled(true)
                    actionBar.setDisplayShowTitleEnabled(true)
                    toolbar.background = ColorDrawable(Color.parseColor("#00407f"))

                    snapHelper?.attachToRecyclerView(boardView.listRecyclerView)
                }

                override fun onAnimationCancel(animation: Animator) {}

                override fun onAnimationRepeat(animation: Animator) {}
            })
            anim.start()
            listViewHolder?.let { trelloBoardAdapter.triggerScaleUpAnimationForListDrag(it, !largeScreen) }

            Toast.makeText(this@BoardActivity, "List moved from col: ${currentDragOperation.startingListIndex} " +
                    "to col: ${currentDragOperation.listIndex}", Toast.LENGTH_LONG).show()
        }

        override fun onListDragEnteredBoardView(boardView: BobBoardView, listViewHolder: BobBoardAdapter.ListViewHolder<BobBoardListAdapter<*>>?, overIndex: Int): Boolean {
            return false
        }

        override fun onListDragExitedBoardView(boardView: BobBoardView, listViewHolder: BobBoardAdapter.ListViewHolder<BobBoardListAdapter<*>>): Boolean {
            return false
        }

        override fun canListDropOver(boardView: BobBoardView, listViewHolder: BobBoardAdapter.ListViewHolder<BobBoardListAdapter<*>>, otherListViewHolder: BobBoardAdapter.ListViewHolder<BobBoardListAdapter<*>>): Boolean {
            if (otherListViewHolder.adapterPosition == board.lists.size) {
                return false
            }
            return true
        }

        override fun canCardDropInList(boardView: BobBoardView, listViewHolder: BobBoardAdapter.ListViewHolder<BobBoardListAdapter<*>>): Boolean {
            if (listViewHolder.itemId == 666L) {
                return false
            }
            return true
        }
    }

    private fun centeredCircleAnimation(view: View, reverse: Boolean): Animator {
        val cx = view.width / 2
        val cy = view.height / 2
        val finalRadius = Math.hypot(cx.toDouble(), cy.toDouble()).toFloat()
        if (reverse) {
            return ViewAnimationUtils.createCircularReveal(view, cx, cy, finalRadius, 0f)
        }
        return ViewAnimationUtils.createCircularReveal(view, cx, cy, 0f, finalRadius)
    }

    fun removeDragListItem() {
        boardView.listRemovedDuringDrag()
        board.lists.removeAt(currentDragOperation.listIndex)
        currentDragOperation.listItem = trelloBoardAdapter.removeItem(currentDragOperation.listIndex)
        currentDragOperation.orphaned = true
    }

    fun insertDragListItem(x: Float, y: Float) {
        val offsetViewBounds = Rect()
        archiveView.getDrawingRect(offsetViewBounds)
        toolbar.offsetDescendantRectToMyCoords(archiveView, offsetViewBounds)
        val destinationIndex = boardView.listAdapterPositionAtPoint(x + offsetViewBounds.left, 0f)
        if (destinationIndex == -1) {
            boardView.listRecyclerView.postOnAnimation {
                insertDragListItem(x, y)
            }
            return
        }
        val index = if (destinationIndex == -1) { currentDragOperation.listIndex } else destinationIndex

        trelloBoardAdapter.insertItem(index, currentDragOperation.listItem!!, true)
        board.lists.add(index, currentDragOperation.listItem!!)
        currentDragOperation.listIndex = index
        currentDragOperation.orphaned = false
        boardView.listRecyclerView.scrollToPosition(index)
    }
}
